### PR TITLE
switch from bundled lazy loading code to the public lazy_loader package

### LINKED ISSUE #489: [FEA] Missing IntelliSense capability in VSCode
**Is your feature request related to a problem? Please describe.**
After installing cucim, unfortunately I cannot get IntelliSense features like auto-complete in Visual Studio Code. According to https://github.com/microsoft/pylance-release/issues/242 this is due to the fact that the Pylance extension does not support IntelliSense for compiled libraries.

**Describe the solution you'd like**
As described in https://github.com/microsoft/pylance-release/issues/242 it would be necessary to provide a type stub ("pyi") file for the library interface.
I would very much appreciate if such a type stub file could be provided.

**Describe alternatives you've considered**
Interestingly, IntelliSense features are working for me in PyCharm. Haven't figured out the reason why PyCharm can auto-generate the type stubs and VSCode can't..

### PR DESCRIPTION
**Pull request by grlee77**
This PR switch from using the internally bundled `_shared/lazy.py` to the public [lazy_loader](https://github.com/scientific-python/lazy_loader) package. 

Also switches to use `.pyi` [stub files](https://mypy.readthedocs.io/en/stable/stubs.html) in cases where submodules or attributes were being dynamically loaded.

closes #489  (probably, but have not yet verified from VS Code)

### CODE REVIEW

**Comment by grlee77** [comment]
rebased on top of #574 which fixes some test failures with the most recent NumPy/CuPy

**Comment by jakirkham** [comment]
Thanks Greg! 🙏

Can you please update [`dependencies.yaml`]( https://github.com/rapidsai/cucim/blob/branch-23.08/dependencies.yaml ) as well?

**Review by gigony — APPROVED** [review]
Thanks @grlee77 ! Looks good to me!

**Comment by jakirkham** [comment]
Looks like there are some conflicts here that need to be resolved

**Comment by jakirkham** [comment]
Thanks Greg! 🙏

Could you please add `lazy_loader >=0.1` to this package list?

https://github.com/rapidsai/cucim/blob/0a6bf42b509408cf26e22d0392dd144ded4eb716/dependencies.yaml#L158-L171

**Comment by jakirkham** [comment]
The CI failure is due to the fact this change needs to be propagated

Please run (in some environment where this dependency can be installed):

```
python -m pip install rapids-dependency-file-generator
rapids-dependency-file-generator
```

Then commit & push the changes

File: python/cucim/src/cucim/skimage/filters/_fft_based.py [hunk 1]
```diff
@@ -159,8 +159,8 @@ def butterworth(
     elif npad > 0:
         center_slice = tuple(slice(npad, s + npad) for s in image.shape)
         image = pad(image, npad, mode='edge')
-    fft_shape = (image.shape if channel_axis is None
-                 else np.delete(image.shape, channel_axis))
+    fft_shape = tuple(image.shape if channel_axis is None
```
[the remark this answers is not in the document]
Comment (grlee77) [reply python/cucim/src/cucim/skimage/filters/_fft_based.py]: This change was merged separately as part of #574 so no longer appears here

File: python/cucim/MANIFEST.in [hunk 2]
```diff
@@ -22,4 +22,6 @@ include tox.ini .travis.yml .appveyor.yml .readthedocs.yml
 include versioneer.py
 include src/cucim/clara/*.so*
 
+recursive-include src/cucim *.pyx *.pxd *.pxi *.py *.pyi *.h *.ini *.npy *.txt *.in *.md
```
Comment (jakirkham) [inline python/cucim/MANIFEST.in:25]: This is a good idea to include. Thanks for adding 🙏

Should we add the `.cpp` files too?

File: python/cucim/setup.py [hunk 3]
```diff
@@ -39,6 +39,7 @@ def read(*names, **kwargs):
     url='https://developer.nvidia.com/multidimensional-image-processing',
     packages=find_packages('src'),
     package_dir={'cucim': 'src/cucim'},
+    package_data={"": ["*.pyi"]},  # distribute
```
Comment (jakirkham) [inline python/cucim/setup.py:42]: Would it make sense to (somewhat?) align this with the `MANIFEST.in`?

**Review by jakirkham — COMMENTED** [review]
Thanks Greg! 🙏

This is a nice simplification

Had a couple questions about files now included in distributions

File: python/cucim/MANIFEST.in (hunk 2 above)
Comment (grlee77) [reply python/cucim/MANIFEST.in:25]: I don't think there are any .cpp files under the Python subfolder, but we do have same .cu files.

Let me add `*.cu` here

Comment (grlee77) [reply python/cucim/MANIFEST.in:25]: could remove `*.pyx *.pxd *.pxi` since we don't have any Cython

File: python/cucim/setup.py (hunk 3 above)
Comment (grlee77) [reply python/cucim/setup.py]: not sure under what conditions it matters (e.g. python setup.py sdist seems to have the .npy files etc that are listed in the MANIFEST.in without listing them again here). I went ahead and added it for consistency, though.

File: python/cucim/MANIFEST.in (hunk 2 above)
Comment (grlee77) [reply python/cucim/MANIFEST.in]: please see latest update

File: python/cucim/setup.py (hunk 3 above)
Comment (jakirkham) [reply python/cucim/setup.py]: Yeah the rule of thumb I try to use is...

* If we want to ship them, but not install them, use `MANIFEST.in`
* If we want to ship and install them, use `package_data`

File: python/cucim/setup.py [hunk 4]
```diff
@@ -39,6 +39,7 @@ def read(*names, **kwargs):
     url='https://developer.nvidia.com/multidimensional-image-processing',
     packages=find_packages('src'),
     package_dir={'cucim': 'src/cucim'},
+    package_data={"": ["*.pyi", "*.h", "*.cu", "*.npy", "*.md", "*.txt"]},
```
Comment (jakirkham) [inline python/cucim/setup.py:42]: Think we can limit this to source code we want to include

Maybe `.npy` if it is used in tests or examples

Text files can probably be dropped

Basically this should include what we want installed in the users environment

```suggestion
    package_data={"": ["*.pyi", "*.h", "*.cu"]},
```

File: python/cucim/MANIFEST.in [hunk 5]
```diff
@@ -22,4 +22,6 @@ include tox.ini .travis.yml .appveyor.yml .readthedocs.yml
 include versioneer.py
 include src/cucim/clara/*.so*
 
+recursive-include src/cucim *.py *.pyi *.cu *.h *.npy *.txt *.in *.md
```
Comment (jakirkham) [inline python/cucim/MANIFEST.in:25]: FWIW `MANIFEST.in` is automatically included. So if that is the only reason for `.in`, we can drop `.in`

```suggestion
recursive-include src/cucim *.py *.pyi *.cu *.h *.npy *.txt *.md
```

File: python/cucim/setup.py (hunk 3 above)
Comment (jakirkham) [reply python/cucim/setup.py]: Think the suggestion in comment ( https://github.com/rapidsai/cucim/pull/575#discussion_r1275479020 ) is revises this to the list we want to install. Though please let me know if we are still missing things

**Review by jakirkham — APPROVED** [review]
Thanks Greg! 🙏

**Comment by jakirkham** [comment]
(asking offline for OPS to review)

**Review by raydouglass — APPROVED** [review]

**Pull request merged**